FoldRegion: some fixes

NoreSoft opened the pull request:
In set_bounds() the local variables start_iter and end_iter are used again. But I'm not sure if the naming is OK.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -253,7 +259,14 @@ gtef_fold_region_get_folded (GtefFoldRegion *fold_region)
-	return priv->folded;
+	if (priv->tag != NULL)
+	{
+		return TRUE;
```
This can be simplified as `return priv->tag != NULL;`

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -277,13 +290,11 @@ gtef_fold_region_set_folded (GtefFoldRegion *fold_region,
 	folded = folded != FALSE;
 
-	if (priv->folded == folded)
+	if ((folded == TRUE && priv->tag != NULL) || (folded == FALSE && priv->tag == NULL))
```
It's better to not duplicate the code, and call the get_folded() function to know the current value.
The variable name can be `currently_folded`, or something like that.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -301,9 +312,8 @@ gtef_fold_region_set_folded (GtefFoldRegion *fold_region,
 		gtk_text_iter_forward_line (&end);
 
-		if (folded)
+		if (folded == TRUE)
```
Booleans should not be compared to TRUE or FALSE. `if (folded)` can be read as an English sentence.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -313,13 +323,19 @@ gtef_fold_region_set_folded (GtefFoldRegion *fold_region,
 						   &end);
+
+			g_object_add_weak_pointer (G_OBJECT (priv->tag),
+						   (gpointer *) &priv->tag);
```
For the tag, a strong ref can be used. The buffer owns the tag, and the buffer will own the FoldRegion, so the FoldRegion can also own the tag (it's not a reference cycle).

swilmet commented:
As a general comment, when fixing stuff, it's better to do small commits (micro commits). So try to fix one FIXME at a time, or a group of related FIXMEs. It's easier to review, and when reading the diff we can be more confident that the change is correct.

NoreSoft commented:
> As a general comment, when fixing stuff, it's better to do small commits (micro commits). So try to fix one FIXME at a time, or a group of related FIXMEs. It's easier to review, and when reading the diff we can be more confident that the change is correct.

You're right, I will keep it in mind. Thank you.

NoreSoft commented:
Pull request updated.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -136,6 +135,11 @@ gtef_fold_region_dispose (GObject *object)
+	if (priv->tag != NULL)
+	{
+		g_clear_object (&priv->tag);
+	}
```
The nice thing about g_clear_object() is that it already checks if `priv->tag != NULL`. So the if condition is not needed.

NoreSoft commented:
On 18.09.2016 13:55, Sébastien Wilmet wrote:
> swilmet requested changes on this pull request.
> 
> Looks good, apart the small thing with g_clear_object().

Cool. :-) I fixed it and updated the pull request.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -126,7 +121,11 @@ gtef_fold_region_dispose (GObject *object)
+		if (priv->tag != NULL)
+		{
+			tag_table = gtk_text_buffer_get_tag_table (priv->buffer);
+			gtk_text_tag_table_remove (tag_table, priv->tag);
```
The tag must be removed from the tag table even if the buffer is already NULL.
A same tag table can be used for several buffers.

NoreSoft replied:
How do we obtain the tag table? Do we store it when storing when creating the text tag?

swilmet-ucl replied:
Yes, do the same as in GtkSourceSearchContext for the found_tag.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -396,29 +396,47 @@ gtef_fold_region_set_bounds (GtefFoldRegion    *fold_region,
+		gtk_text_iter_forward_line (&start_iter);
+		gtk_text_iter_forward_line (&end_iter);
+
+		gtk_text_buffer_remove_tag (priv->buffer, priv->tag, &start_iter, &end_iter);
```
It's not robust to remove the tag this way. The tag is used only by a single FoldRegion. So it's more robust to remove completely the tag from the buffer, and create another tag.

NoreSoft replied:
Why isn't it robust?

swilmet-ucl replied:
The buffer can be modified. It is better to remove the tag completely from the buffer, wherever it is applied. And with a reference to the buffer and the tag, you exactly know where the tag is applied (by navigating with the `gtk_text_iter_forward_to_tag_toggle()` function). But here since you need to remove the tag, it is easier to just destroy the tag and create a new one (with separate functions to do that, so you can call the same static function when you need to apply the tag).

NoreSoft commented:
Pull request updated.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -41,21 +41,43 @@ struct _GtefFoldRegionPrivate
 {
 	GtkTextBuffer *buffer;
 	GtkTextTag *tag;
+	GtkTextTagTable *tag_table;
```
In GtkSourceSearchContext there is a comment in the struct explaining why we need the `tag_table`. I think it's useful information, when trying to understand some code, the data is the most important (the code can be deduced from the data structures). So it is important to well document the structs (document the invariants, why a field is needed, why such a data structure is used and not another one (if it isn't obvious), etc).

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -41,21 +41,43 @@ struct _GtefFoldRegionPrivate
 G_DEFINE_TYPE_WITH_PRIVATE (GtefFoldRegion, gtef_fold_region, G_TYPE_OBJECT)
 
 static void
+gtef_fold_region_create_tag (GtefFoldRegion *fold_region)
```
static functions don't need to have the namespace and classname, except when overriding a vfunc (that's the exception). So here it can be just `create_tag()`.

swilmet commented on gtef/gtef-fold-region.c:
```diff
@@ -41,21 +41,43 @@ struct _GtefFoldRegionPrivate
+{
+	GtefFoldRegionPrivate *priv = gtef_fold_region_get_instance_private (fold_region);
+
+	priv->tag = gtk_text_buffer_create_tag (priv->buffer,
```
the function doesn't check if priv->tag already exists. Either add a `g_assert()`, or call a function to remove the tag before creating a new one.

NoreSoft commented:
Pull request updated.

swilmet commented:
Ok, it was easier for me to merge the patch, but I've added another set of FIXMEs.
Look at the diff of my commit, I changed a little the code besides adding the FIXMEs.
